pkg/search/postgres/store_cache.go

```diff
@@ -294,13 +295,28 @@ func (c *cachedStore[T, PT]) GetMany(ctx context.Context, identifiers []string)
 }
 
 // WalkByQuery iterates over all the objects scoped by the query applies the closure.
-func (c *cachedStore[T, PT]) WalkByQuery(ctx context.Context, q *v1.Query, fn func(obj PT) error) error {
-	if q == nil || protocompat.Equal(q, search.EmptyQuery()) {
-		c.cacheLock.RLock()
-		defer c.cacheLock.RUnlock()
-		return c.walkCacheNoLock(ctx, fn)
+func (c *cachedStore[T, PT]) WalkByQuery(ctx context.Context, query *v1.Query, fn func(obj PT) error) error {
+	c.cacheLock.RLock()
+	defer c.cacheLock.RUnlock()
+	identifiers, err := c.underlyingStore.GetIDsByQuery(ctx, query)
+	// Fallback to the underlying store on error.
+	if err != nil {
+		log.Errorf("Failed to get identifiers by query, falling back to walk results by query: %v", err)
+		return c.underlyingStore.WalkByQuery(ctx, query, fn)
+	}
+	for _, id := range identifiers {
+		obj, found := c.cache[id]
+		if !found {
+			return errox.InvariantViolation.Newf("object %q not found in store cache", id)
```

Comment:
If I remember correctly, the lock at cache level is only acquired after the DB call has returned. This leaves a short window of possible desynchronization between cache and store.
I'd rather have this desync condition as Warn log than error, but can go with the condition being an error.